classes/observers/listeners/UsageEventLog.inc.php

```diff
@@ -0,0 +1,472 @@
+                $countryIsoCode = $geoIPRecord->country->isoCode;
+                // When found, up to three characters long subdivision portion of the ISO 3166-2 code is returned
+                // s. https://github.com/maxmind/GeoIP2-php/blob/main/src/Record/Subdivision.php#L20
+                $regionIsoCode = $geoIPRecord->mostSpecificSubdivision->isoCode;
+                // DB-IP IP to City Lite database does not provide region Iso code but name,
+                // thus try to get the region Iso code by the name,
+                // but we need country for that
+                if (!isset($regionIsoCode) && isset($countryIsoCode)) {
+                    $regionName = $geoIPRecord->mostSpecificSubdivision->name;
+                    if (isset($regionName)) {
+                        $isoCodes = app(IsoCodesFactory::class);
+                        $allCountryRegions = $isoCodes->getSubdivisions()->getAllByCountryCode($countryIsoCode);
+                        foreach ($allCountryRegions as $countryRegion) {
+                            if ($countryRegion->getName() == $regionName) {
+                                $regionIsoCodeArray = explode('-', $countryRegion->getCode());
+                                $regionIsoCode = $regionIsoCodeArray[1];
+                                break;
+                            }
+                        }
+                    }
+                }
+                $cityName = $geoIPRecord->city->name;
+            } catch (\BadMethodCallException $e) {
+                error_log('There was a problem using city method on the Geo database at ' . $this->getGeoDBPath() . '. Error: ' . $e->getMessage());
+            } catch (\GeoIp2\Exception\AddressNotFoundException $e) {
+                error_log('There was a problem finding IP in the Geo database at ' . $this->getGeoDBPath() . '. Error: ' . $e->getMessage());
+            } catch (\MaxMind\Db\Reader\InvalidDatabaseException $e) {
+                error_log('There was a problem reading the Geo database at ' . $this->getGeoDBPath() . '. Error: ' . $e->getMessage());
+            }
+        }
+        $cachedGeoData[$hashedIp]['country'] = $countryIsoCode;
+        $cachedGeoData[$hashedIp]['region'] = $regionIsoCode;
+        $cachedGeoData[$hashedIp]['city'] = $cityName;
+        $this->geoDataCache->setEntireCache($cachedGeoData);
+        return $cachedGeoData[$hashedIp];
+    }
+
+    /**
+    * Geo cache miss callback.
+    */
+    public function geoDataCacheMiss(FileCache $cache): array
+    {
+        $cache->setEntireCache([]);
+        return [];
+    }
+
+    /**
+    * Get document type based on the mimetype
+    * The mimetypes considered here are subset of those used in PKPFileService::getDocumentType()
+    *
+    * @return int One of the StatisticsHelper::STATISTICS_FILE_TYPE_ constants
+    */
+    protected function getDocumentType(string $mimetype): int
+    {
+        switch ($mimetype) {
```

Comment:
The MIME type detection tools often don't work well on certain content types (e.g. HTML), so `PKPString::mime_content_type` adds some additional mechanisms to use the file extension in cases where the MIME type is frequently misdetected. Both OpenOffice/LibreOffice and HTML are frequently misdetected types. I suspect Flysystem (which we're using lately for filesystem abstraction) improves on that -- I'd suggest checking over https://flysystem.thephpleague.com/docs/advanced/mime-type-detection/#article to see.

Reply:
I am not sure but: this function already gets the calculated/detected mimetype -- it only then returns the COUNTER file type depending on the mime type. So do you think I should do something else here?

Reply:
As coded, it will frequently detect `text/xml` instead of `text/html`, and `application/zip` instead of the `application/vnd.openxmlformats-officedocument.*` MIME types.

Reply:
Ah! OK, then you should be fine, thanks.